core/internal/featurechecker/features.go

```diff
@@ -0,0 +1,115 @@
+package featurechecker
+
+import (
+	"context"
+	"sync"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/wandb/wandb/core/internal/gql"
+	"github.com/wandb/wandb/core/internal/observability"
+	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
+)
+
+// ServerFeatures is responsible for providing the capabilities of a server
+// Features and capabilities are retrieved from the server via a GraphQL query
+type ServerFeatures struct {
+	Features map[spb.ServerFeature]ServerFeature
+
+	ctx           context.Context
+	graphqlClient graphql.Client
+	mu            sync.Mutex
+}
+
+type ServerFeature struct {
+	DefaultValue string
+	Description  string
+	Enabled      bool
+	Name         string
+}
+
+func NewServerFeatures(
+	ctx context.Context,
+	graphqlClient graphql.Client,
+) *ServerFeatures {
+	var features map[spb.ServerFeature]ServerFeature
+
+	// If graphqlClient is nil, we won't ever be able to query the server
+	// So create an empty map which will default to false for all features when a key is not found
+	if graphqlClient == nil {
+		features = make(map[spb.ServerFeature]ServerFeature)
+	}
+
+	return &ServerFeatures{
+		Features:      features,
+		ctx:           ctx,
+		graphqlClient: graphqlClient,
+		mu:            sync.Mutex{},
+	}
+}
+
+func (sf *ServerFeatures) getServerFeatures() {
+	sf.Features = map[spb.ServerFeature]ServerFeature{}
+
+	// Query the server for the features provided by the server
+	resp, err := gql.ServerFeaturesQuery(
+		sf.ctx,
+		sf.graphqlClient,
+		gql.RampIDTypeUsername,
+	)
+	if err != nil {
+		return
+	}
+
+	for _, f := range resp.ServerInfo.FeatureFlags {
+		featureName := spb.ServerFeature(spb.ServerFeature_value[f.RampKey])
+		sf.Features[featureName] = ServerFeature{
+			Name:    f.RampKey,
+			Enabled: f.IsEnabled,
+		}
+	}
+}
+
```

Comment:
you could use this signature:
```
func (sf *ServerFeatures) GetFeatures(names ...spb.ServerFeature)
```

this way you can implement it once for any number of names

Reply:
I think it still might make sense to have a GetFeature for a single feature. This way we can avoid creating a new map if only one feature value is retrieved, thoughts?

Reply:
it depends on the usage, would be good to have in this PR examples of expected usage, if we only use it to get the protobuf message then it doesn't seem like we need the single one, but if we use it for something else maybe there is a value there

Reply:
Fair point, will add some examples to the overview of this PR.

---

A single `GetFeature` function will be useful for getting feature values from within the Go code (our long term vision) Since in almost every case we will be checking the value directly, rather than checking multiple values at the same time.

example in go:

```
newFeatureIsEnabled := featureChecker.GetFeature(newFeatureName).Enabled
if (newFeatureIsEnabled) {
  // do something
}
```

while if we have only a 'get one or more features` function it adds extra time for creating a map around one or more features and then we need to index on the map to get the feature we want
example:
```
newFeatureIsEnabled := featureChecker.GetFeatures(newFeatureName)[newFeatureName].Enabled
if (newFeatureIsEnabled) {
  // do something
}
```

Reply:
In fact now that I am thinking more on this, I think we should just simplify this for now and only allow getting a single feature at once (even for the python client)

right now there is not a use case for getting multiple features, and this can be added later if needed.

Reply:
yeah if we have this use-case i agree it makes sense to exactly extract it and not returning a dict!